File: app/src/main/java/com/tianma/xsmscode/ui/block/AppBlockFragment.java
package com.tianma.xsmscode.ui.block;

import android.app.Activity;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.github.tianma8023.xposed.smscode.R;
import com.tianma.xsmscode.common.adapter.ItemCallback;
import com.tianma.xsmscode.common.utils.SnackbarHelper;
import com.tianma.xsmscode.data.db.entity.AppInfo;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import com.github.tianma8023.xposed.smscode.databinding.FragmentAppBlockBinding;
import dagger.android.support.DaggerFragment;

public class AppBlockFragment extends DaggerFragment implements AppBlockContract.View {

    private FragmentAppBlockBinding binding;

    private AppInfoAdapter mAppInfoAdapter;

    private Activity mActivity;

    @Inject
    AppBlockContract.Presenter mPresenter;

    static AppBlockFragment newInstance() {
        return new AppBlockFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        binding = FragmentAppBlockBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mActivity = getActivity();

        mAppInfoAdapter = new AppInfoAdapter(mActivity, new ArrayList<>());
        mAppInfoAdapter.setItemCallback(new ItemCallback<AppInfo>() {
            @Override
            public void onItemClicked(View itemView, AppInfo item, int position) {
                itemClicked(item, position);
            }

            @Override
            public boolean onItemLongClicked(View itemView, AppInfo item, int position) {
                itemClicked(item, position);
                return true;
            }

            @Override
            public void onCreateItemContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo, AppInfo item, int position) {

            }
        });

        binding.appBlockRecyclerView.setLayoutManager(new LinearLayoutManager(mActivity));
        binding.appBlockRecyclerView.setAdapter(mAppInfoAdapter);
        binding.appBlockRecyclerView.addItemDecoration(new DividerItemDecoration(mActivity, DividerItemDecoration.VERTICAL));

        binding.swipeRefreshLayout.setOnRefreshListener(() -> mPresenter.refreshData());
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.refreshData();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mPresenter.onDetach();
        cancelProgress();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_app_block, menu);

        MenuItem searchItem = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) searchItem.getActionView();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                doFilter(query);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                doFilter(newText);
                return true;
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_sort_by_label_asc) {
            mPresenter.doSort(SortType.LABEL_ASC);
        } else if (id == R.id.action_sort_by_pkg_asc) {
            mPresenter.doSort(SortType.PACKAGE_ASC);
        } else if (id == R.id.action_sort_by_label_desc) {
            mPresenter.doSort(SortType.LABEL_DESC);
        } else if (id == R.id.action_sort_by_pkg_desc) {
            mPresenter.doSort(SortType.PACKAGE_DESC);
        } else if (id == R.id.action_tick) {
            mPresenter.saveData();
        } else {
            return super.onOptionsItemSelected(item);
        }
        return true;
    }

    private void doFilter(String filter) {
        mPresenter.doFilter(filter);
    }

    private void itemClicked(AppInfo appInfo, int position) {
        mPresenter.doItemClicked(appInfo);
        mAppInfoAdapter.setItemSelected(position);
    }

    @Override
    public void showData(List<AppInfo> appInfoList) {
        mAppInfoAdapter.setItemList(appInfoList);
        binding.swipeRefreshLayout.setEnabled(false);
    }

    @Override
    public void showError(Throwable t) {
        binding.swipeRefreshLayout.setEnabled(true);
        SnackbarHelper.makeShort(binding.appBlockRecyclerView, R.string.load_failed).show();
    }

    @Override
    public void showProgress() {
        if (!binding.swipeRefreshLayout.isRefreshing()) {
            binding.swipeRefreshLayout.setRefreshing(true);
        }
    }

    @Override
    public void cancelProgress() {
        if (binding.swipeRefreshLayout.isRefreshing()) {
            binding.swipeRefreshLayout.setRefreshing(false);
        }
    }

    @Override
    public void onSaveSuccess() {
        mActivity.onBackPressed();
    }

    @Override
    public void onSaveFailed() {
        SnackbarHelper.makeShort(binding.appBlockRecyclerView, R.string.save_failed).show();
    }
}

File: app/src/main/java/com/tianma/xsmscode/ui/block/AppBlockFragment.kt
package com.tianma.xsmscode.ui.block

import android.app.Activity
import android.os.Bundle
import android.view.ContextMenu
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.RecyclerView
import com.github.tianma8023.xposed.smscode.R
import com.github.tianma8023.xposed.smscode.databinding.FragmentAppBlockBinding
import com.tianma.xsmscode.common.adapter.ItemCallback
import com.tianma.xsmscode.common.utils.SnackbarHelper
import com.tianma.xsmscode.data.db.entity.AppInfo
import java.util.ArrayList

class AppBlockFragment : Fragment() {

    private var _binding: FragmentAppBlockBinding? = null
    private val binding get() = _binding!!

    private var mAppInfoAdapter: AppInfoAdapter? = null
    private var mActivity: Activity? = null
    private lateinit var mViewModel: AppBlockViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentAppBlockBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mActivity = requireActivity()
        mViewModel = ViewModelProvider(this).get(AppBlockViewModel::class.java)
        setupObservers()

        mAppInfoAdapter = AppInfoAdapter(requireContext(), ArrayList())
        mAppInfoAdapter?.setItemCallback(object : ItemCallback<AppInfo> {
            override fun onItemClicked(itemView: View, item: AppInfo, position: Int) {
                itemClicked(item, position)
            }

            override fun onItemLongClicked(itemView: View, item: AppInfo, position: Int): Boolean {
                itemClicked(item, position)
                return true
            }

            override fun onCreateItemContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo?, item: AppInfo, position: Int) {
            }
        })

        binding.appBlockRecyclerView.layoutManager = LinearLayoutManager(mActivity)
        binding.appBlockRecyclerView.adapter = mAppInfoAdapter
        binding.appBlockRecyclerView.addItemDecoration(DividerItemDecoration(mActivity, DividerItemDecoration.VERTICAL))

        binding.swipeRefreshLayout.setOnRefreshListener { mViewModel.refreshData() }
        
        mViewModel.refreshData()

        setupMenu()
    }

    private fun setupMenu() {
        requireActivity().addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_app_block, menu)

                val searchItem = menu.findItem(R.id.action_search)
                val searchView = searchItem.actionView as SearchView
                searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                    override fun onQueryTextSubmit(query: String): Boolean {
                        doFilter(query)
                        return true
                    }

                    override fun onQueryTextChange(newText: String): Boolean {
                        doFilter(newText)
                        return true
                    }
                })
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                when (menuItem.itemId) {
                    R.id.action_sort_by_label_asc -> mViewModel.doSort(SortType.LABEL_ASC)
                    R.id.action_sort_by_pkg_asc -> mViewModel.doSort(SortType.PACKAGE_ASC)
                    R.id.action_sort_by_label_desc -> mViewModel.doSort(SortType.LABEL_DESC)
                    R.id.action_sort_by_pkg_desc -> mViewModel.doSort(SortType.PACKAGE_DESC)
                    R.id.action_tick -> mViewModel.saveData()
                    else -> return false
                }
                return true
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    private fun setupObservers() {
        mViewModel.appsLiveData.observe(viewLifecycleOwner) { showData(it) }
        mViewModel.loadingLiveData.observe(viewLifecycleOwner) { isLoading ->
            if (isLoading) {
                showProgress()
            } else {
                cancelProgress()
            }
        }
        mViewModel.errorEvent.observe(viewLifecycleOwner) { showError(it) }
        mViewModel.saveSuccessEvent.observe(viewLifecycleOwner) { onSaveSuccess() }
        mViewModel.saveFailedEvent.observe(viewLifecycleOwner) { onSaveFailed() }
    }

    override fun onDestroy() {
        super.onDestroy()
        cancelProgress()
    }


    private fun doFilter(filter: String) {
        mViewModel.doFilter(filter)
    }

    private fun itemClicked(appInfo: AppInfo, position: Int) {
        mViewModel.doItemClicked(appInfo)
        mAppInfoAdapter?.setItemSelected(position)
    }

    private fun showData(appInfoList: List<AppInfo>) {
        mAppInfoAdapter?.setItemList(appInfoList)
        binding.swipeRefreshLayout.isEnabled = false
    }

    private fun showError(t: Throwable) {
        binding.swipeRefreshLayout.isEnabled = true
        SnackbarHelper.makeShort(binding.appBlockRecyclerView, R.string.load_failed).show()
    }

    private fun showProgress() {
        if (!binding.swipeRefreshLayout.isRefreshing) {
            binding.swipeRefreshLayout.isRefreshing = true
        }
    }

    private fun cancelProgress() {
        if (binding.swipeRefreshLayout.isRefreshing) {
            binding.swipeRefreshLayout.isRefreshing = false
        }
    }

    private fun onSaveSuccess() {
        requireActivity().onBackPressedDispatcher.onBackPressed()
    }

    private fun onSaveFailed() {
        SnackbarHelper.makeShort(binding.appBlockRecyclerView, R.string.save_failed).show()
    }

    companion object {
        fun newInstance(): AppBlockFragment {
            return AppBlockFragment()
        }
    }
}
